xmipptomo/protocols/protocol_extract_subtomos.py

```diff
@@ -0,0 +1,222 @@
+# **************************************************************************
+# *
+# * Authors:     J.L. Vilas (jlvilas@cnb.csic.es)
+# *
+# * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
+# *
+# * This program is free software; you can redistribute it and/or modify
+# * it under the terms of the GNU General Public License as published by
+# * the Free Software Foundation; either version 2 of the License, or
+# * (at your option) any later version.
+# *
+# * This program is distributed in the hope that it will be useful,
+# * but WITHOUT ANY WARRANTY; without even the implied warranty of
+# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# * GNU General Public License for more details.
+# *
+# * You should have received a copy of the GNU General Public License
+# * along with this program; if not, write to the Free Software
+# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+# * 02111-1307  USA
+# *
+# *  All comments concerning this program package may be sent to the
+# *  e-mail address 'scipion@cnb.csic.es'
+# *
+# **************************************************************************
+
+import os
+import glob
+from pwem.emlib import lib
+from pwem.objects import Particle, Volume, Transform, String, SetOfVolumes, SetOfParticles
+from pwem.protocols import EMProtocol
+from pwem import ALIGN_PROJ
+import pwem.emlib.metadata as md
+
+from pyworkflow import BETA
+from pyworkflow import utils as pwutils
+from pyworkflow.protocol.params import PointerParam, EnumParam, IntParam, BooleanParam
+
+from tomo.objects import SetOfTomograms, SetOfSubTomograms, SubTomogram, SetOfCoordinates3D, TomoAcquisition, MATRIX_CONVERSION
+import tomo.constants as const
+
+from tomo.protocols import ProtTomoBase
+from xmipp3.convert import alignmentToRow
+
+
+COORD_BASE_FN = 'coords'
+# Tomogram type constants for particle extraction
+SAME_AS_PICKING = 0
+OTHER = 1
+
+
+class XmippProtExtractSubtomos(EMProtocol, ProtTomoBase):
+    """
+    Extract a set of subtomograms from a set of tomograms given a set of coordinates.
+    """
+    _label = 'extract subtomos'
+    _devStatus = BETA
+    #_possibleOutputs = SetOfSubTomograms
+    #OUTPUT_PREFIX = _possibleOutputs.subtomograms.name
+    lines = []
+    tomoFiles = []
+
+    # --------------------------- DEFINE param functions ------------------------
+    def _defineParams(self, form):
+        form.addSection(label='Parameters')
+
+        form.addParam('tomograms', PointerParam, pointerClass=SetOfTomograms,
```

Comment:
Since coordinates have a set of tomograms associated I'd ask for 3d coordinates FIRST, , and optionally, a setOfTomograms